platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,852 @@
+      but may be subject to change later if we can figure out a better way
+- Render the form navigation buttons "Back" and "Continue"
+- Prevent navigation if there are validation errors
+
+##### `Route` Component
+**Purpose:**
+The `Route` will be re-exported from `react-router-dom`. The purpose of
+re-exporting is twofold:
+- Simplifies imports
+- Ensures only one version of `react-router-dom` is used
+
+**Usage requirements:**
+- `Route` may only be a child of `Router`
+
+**Why not just use `Page`?**
+The `Chapter` and `Page` components are meant to be used to define pages where
+the user may enter data. These pages are sequential and have back and continue
+buttons. `Route`s are used for pages in the application that live outside that
+form flow. These pages may include:
+- `/introduction`
+- `/form-saved`
+- `/error`
+
+##### Functionality: Array pages
+Array pages are one or more pages that are looped through in a form flow, once
+per item in the array. To help understand the nuances of array pages, consider
+the following example:
+
+```jsx
+<Page
+  path="/marriages/:marriageIndex/foo"
+  pathParams={{
+    marriageIndex: 'marriages'
+  }}>
+  {({ formData, pathParams }) => (
+    <FieldArray
+      name="marriages"
+      render={arrayHelpers => (
+        <Field name={`marriages.${pathParams.marriageIndex}.name.first`} />
+      )}
+    />
+  )}
+</Page>
+```
+
+For array pages to work we need a couple core things:
+- The path for the page
+- The data for the current item in the array
+
+To get these pieces, `Page` will accept the `path` and `pathParams` props.
+
+`path` will be the URL within the application and must (for array pages) contain
+URL parameters (term pending). URL parameters must be the only token between
+`/`s or at the end of the URL. For example, `/:index/foo` and `/:index` are
+valid path parameters. `/my-:index-item` is an invalid path parameter.
+
+`pathParams` will be an object whose keys map to the path parameters and whose
+values map to the path to the form data. In the above example, the `pathParams`
+will assign `:marriageIndex` in the URL a number based in the current item in
+the `marriages` array.
+
+##### Review page
+To keep Formulate simple, it will not build the review page automatically. It
+will be the responsibility of the engineers to build the review page. This
+decision is subject to change, but as of the time of this writing, I can't think
+of a way to automatically build the review page and meet all of the following
+requirements:
+- The chapters are divided correctly
+- The pages are divided correctly
+- The edit mode for each page is identical to the page in the form flow
+- The review mode for each page displays each field on the page in a data table
+  with the appropriate label and value
+- Each field in the review mode may be overridden when the label or value need
+  to be changed for a better user experience
+- Each page may or may not have a review mode
```

Comment:
This may not be the right place, but I'm curious to learn more about this even though I've known about it for a while. Do solutions for some of these requirements complicate the solution for others?

Reply:
:thinking: Maybe?

In the context of this document, the struggle I had was doing this automatically when there is no easily digestible single source of truth for all pages. In the schema-based forms library, we had access to all the configuration needed to build all the fields on all the pages. With this new approach, that data lives in the props passed to the fields, so we'd have to somehow start at `<Router>` and iterate through all its children to get all the chapters and pages, all their fields, and build out the review mode for each page, making lots of assumptions along the way, _and_ then provide hooks in for engineers to override the automatically-built stuff when it's wrong for whatever reason.

So at least here, the problem wasn't that the solution to some of the problems complicated the solution to other problems (thought that's probably true). The problem was more just doing it at all. :sweat_smile: